colexec: optimize resetting of buffered groups in merge joiner

yuzefovich opened the pull request:
**colexec: fix resetting of buffered groups**

Release justification: bug fixes and low-risk updates to new
functionality.

Previously, whenever we needed to reset the buffered groups, we would
close the spilling queue and would create a new one when needed. This is
an overkill since we could simply reset the spilling queues that we have
which reduces amount of allocations and improves the performance.

Addresses: #46502.

Release note: None

**logictest: make results of queries in vectorize test deterministic**

Release justification: non-production code changes.

Several queries in `vectorize` logic test could have produced different
results, and this is now fixed.

Fixes: #46630.

Release note: None

yuzefovich committed:
colexec: fix resetting of buffered groups

Release justification: bug fixes and low-risk updates to new
functionality.

Previously, whenever we needed to reset the buffered groups, we would
close the spilling queue and would create a new one when needed. This is
an overkill since we could simply reset the spilling queues that we have
which reduces amount of allocations and improves the performance.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46587)
<!-- Reviewable:end -->

yuzefovich commented:
Before:
![Screen Shot 2020-03-25 at 1 41 18 PM](https://user-images.githubusercontent.com/5135385/77583545-7e3c7180-6e9e-11ea-8495-c1748bd5a356.png)
After:
![Screen Shot 2020-03-25 at 1 41 23 PM](https://user-images.githubusercontent.com/5135385/77583553-81cff880-6e9e-11ea-8aae-d3b6b494aaf1.png)
Note that the benchmark still shows a lot of allocations (for line `items: make([]coldata.Batch, itemsLen),`) because we're creating a new merge join operator in each run, but in the real world these allocations would not matter much.

This gives us 10-20% improvement on the benchmark, but we're still slower by 30-40% than in 19.2. Yet in absolute values the speed measured is like 1200MB/s, so merge joiner should never be the bottleneck of the query, plus we added memory accounting and disk spilling support, so the performance hit is expected and acceptable.

yuzefovich committed:
logictest: make results of queries in vectorize test deterministic

Release justification: non-production code changes.

Several queries in `vectorize` logic test could have produced different
results, and this is now fixed.

Release note: None

asubiotto reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46587)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/mergejoiner.go, line 446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46587#-M3LHsEYEwFTWGoOBU6O:-M3LHsEYEwFTWGoOBU6P:b-6cud5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e297981537778db177b4a252f790109ba630b511/pkg/sql/colexec/mergejoiner.go#L446)):*
> ```Go
> 	o.proberState.lBufferedGroup.spillingQueue = newSpillingQueue(
> 		o.unlimitedAllocator, o.left.sourceTypes, o.memoryLimit,
> 		o.diskQueueCfg, o.fdSemaphore, coldata.BatchSize(), o.diskAcc,
> ```

`s/coldata.BatchSize/outBatchSize`? This might help benchmarks as we could be prematurely spilling since the items slice is sized accordingly. Smaller batches mean a larger items slice.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46587)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/mergejoiner.go, line 446 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46587#-M3LHsEYEwFTWGoOBU6O:-M3N5YVbBU1vE79ZPENH:bazubd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e297981537778db177b4a252f790109ba630b511/pkg/sql/colexec/mergejoiner.go#L446)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`s/coldata.BatchSize/outBatchSize`? This might help benchmarks as we could be prematurely spilling since the items slice is sized accordingly. Smaller batches mean a larger items slice.
</blockquote></details>

`outBatchSize` matters only for the output batches of the merge joiner. The batches that are enqueued to `spillingQueue` can be at most `coldata.BatchSize()` (i.e. at most the input batch size, not the output). Also, I don't think in benchmarks we actually spill to disk, ever, so this would not help.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46587#-:-M3RYMXS2SwkfRWOp4gx:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

yuzefovich commented:
bors r+
